text/0000-list-rendering-reform.md

```diff
@@ -0,0 +1,306 @@
+
+```html
+<template lwc:each={outer} lwc:key={item} lwc:item="item" lwc:index="index">
+  <!-- These will be the outer loop variables -->
+  Outer item: {item}
+  Outer index: {index}
+  <template lwc:each={inner} lwc:key={item} lwc:item="item" lwc:index="index">
+    <!-- These will be the inner loop variables -->
+    Inner item: {item}
+    Inner index: {index}
+  </template>
+</template>
+```
+
+Note that the two loops have the same variable names for `lwc:item` and `lwc:index`, and this does not cause a compiler error.
+
+Also note that the value of the `lwc:key` in each case refers to the `item` on the same `<template>` element.
+
+#### Empty values
+
+Empty values (e.g. `lwc:index=""`) will also throw a compile-time error. For example:
+
+```html
+<template lwc:each={items} lwc:key={item.id} lwc:item="item" lwc:index="">
+</template>
+```
+
+Missing values (e.g. `lwc:index`) will also throw at compile-time:
+
+```html
+<template lwc:each={items} lwc:key={item.id} lwc:item="item" lwc:index>
+</template>
+```
+
+#### Combining with `for:each` and `iterator:*`
+
+All new directives in this RFC must _not_ be combined in the same template file with any directive associated with the
+existing `for:each` or `iterator:*` directives. Doing so should throw a compile-time error.
+
+New directives: `lwc:each`, `lwc:key`, `lwc:item`, `lwc:index`, `lwc:first`, `lwc:last`, `lwc:even`, `lwc:odd`.
+
+Old directives: `for:each`, `for:item`, `for:index`, `key`, `iterator:*`.
+
+If any of the directives from the first group are in a template file with any directive from the second group, then a compile-time error is thrown.
+
+The reason for this is that mixing the new and old directives could easily lead to confusion, e.g. when `lwc:key` and `key` are mixed in the same file.
+
+#### Other constraints
+
+The directives may appear in any order on the element.
+
+Any other constraints that apply to `for:each` must also apply to `lwc:each`. For example, `lwc:ref` is (currently) unsupported inside of `for:each` and `iterator:*` – the same applies to `lwc:each`. For another example, `<slot>`s cannot have any of the new directives, and cannot be repeated inside the new iterator.
+
+## Drawbacks
+
+This new directive may cause confusion, because we are introducing a third directive to try to unify the previous two (see [XKCD 927](https://xkcd.com/927/)).
+
+The implementation also becomes more complex, as we will need to support all three directives for the foreseeable future. Perhaps they can share logic under the hood, but we will still need to test all three, at the very least.
+
+We will also need to make an effort to advocate for adoption of the new directive, which may involve updating existing documentation and writing codemods or IDE tooling.

```

Comment:
Any thoughts about adding compiler warnings?

Reply:
I think similar to `lwc:if`/`lwc:else`, it may be a bit early to add compiler warnings. Compiler warnings to me should (normally) be reserved for truly broken things (like using a `key` outside of an iteration, or invalid HTML syntax that is allowed by old parse5 versions). In the future, though – why not?